Add ERC: Sovereign Bridged Token (#89)
ethereum/ERCs: ERCS/eip-7281.md

Reviewer: function setLimits(address _bridge, uint256 _mintingLimit, uint256 _burningLimit) external;
  
  pls, change it to `function setBridgeLimits(...` to distinguish that from other ERC20 features / extensions.

---
eip: 7281
title: Sovereign Bridged Token
description: An interface for creating fungible representations of tokens bridged across domains.
author: Shaito (@0xShaito), Excalibor (@excaliborr), Arjun Bhuptani (@arjunbhuptani)
discussions-to: https://ethereum-magicians.org/t/erc-7281-sovereign-bridged-tokens/14979
status: Draft
type: Standards Track
category: ERC
created: 2023-06-27
requires: 20
---

## Abstract

> This proposal defines a minimal extension to [ERC-20](../EIPS/eip-20.md) (affectionately called xERC-20) that enables bridging tokens across domains without creating multiple infungible representations of the same underlying asset. Token issuers can define which bridges are able to mint or burn the xERC-20 on a given domain and configure rate limits for each allowed bridge.

Reviewer 2: The bots will complain a lot about "xERC-20" just because of how close it is to a valid "ERC-XXXX" identifier. You could go with `XERC20` as your interface name, then use \`XERC20\` (note the backticks) when referring to it.

> This proposal defines a minimal extension to [ERC-20](../EIPS/eip-20.md) (affectionately called xERC-20) that enables bridging tokens across domains without creating multiple infungible representations of the same underlying asset. Token issuers can define which bridges are able to mint or burn the xERC-20 on a given domain and configure rate limits for each allowed bridge.

Reviewer 2: ```suggestion
This proposal defines a minimal extension to [ERC-20](./eip-20.md) (affectionately called xERC-20) that enables bridging tokens across domains without creating multiple infungible representations of the same underlying asset. Token issuers can define which bridges are able to mint or burn the xERC-20 on a given domain and configure rate limits for each allowed bridge.
```

- Maker’s Teleport facility allows for minting and burning canonical DAI between domains. While the approach solves for the desirable properties above, it is highly custom to Maker’s architecture and relies on Maker’s own economic security to function. Circle CCTP similarly solves this problem, but using a mechanism that only centralized token issuers can implement.
- Token issuer multi-bridge implementations, e.g. Angle protocol, Frax Ferry, and Threshold Network’s tBTC. These examples solve for some or all of the above desiderata and can be applied more broadly to all tokens if coupled with minor additions for compatibility with existing deployed tokens.

## Specification

The key words "MUST", "MUST NOT", "REQUIRED", "SHALL", "SHALL NOT", "SHOULD", "SHOULD NOT", "RECOMMENDED", "NOT RECOMMENDED", "MAY", and "OPTIONAL" in this document are to be interpreted as described in RFC 2119 and RFC 8174.

### Overview

The system proposed below has two main components:

- A standard interface for bridged tokens
- A wrapper (aka “Lockbox”) that allows existing tokens to adopt the above interface

### Token Interface

> All xERC-20 tokens MUST implement the standard ERC-20 interface.  Note that while many of the below functions are inspired by ERC-777, implementers are NOT REQUIRED to adhere to the full ERC-777 specification.

Reviewer 2: ```suggestion
All xERC-20 tokens MUST implement the standard ERC-20 interface.  Note that while many of the below functions are inspired by [ERC-777](./eip-777.md), implementers are NOT REQUIRED to adhere to the full ERC-777 specification.
```

> All EIP-XX tokens MUST implement the following interface.

Reviewer 2: If you use HTML-style comments, the linter will make sure you fill in placeholders:

```suggestion
All EIP-XX <!-- TODO --> tokens MUST implement the following interface.
```

```ts

> interface IXERC-20 {

Reviewer 2: ```suggestion
interface IXERC20 {
```

### Fungibility

The core problem that this standard tackles is the **fungibility** of tokens bridged across domains against the “canonical” token on that domain. (Note: which token is “canonical” is dictated by the token issuer and is sometimes, but not always, the token minted by a given domain’s enshrined bridge - e.g. a rollup bridge).

![](../assets/eip-draft_bridged_tokens/fragmentation.png)

It is not enough to simply allow multiple bridges to mint the same representation on a remote domain if liquidity is locked into bridges on the home domain. To illustrate this, consider an example where two bridges control minting rights of canonical USDT on an L2:

- Alice bridges 100 USDT from L1→L2 through bridge 1. The underlying L1 USDT tokens is locked in bridge 1 and 100 USDT is minted on L2.
- Bob bridges 100 USDT from L1→L2 through bridge 2. Similar to the above case, the underlying L1 USDT tokens are locked in bridge 2 and 100 USDT is minted on L2.
- Suppose Alice pays Bob her 100 USDT. Bob now has 200 USDT.
- Bob attempts to bridge the 200 USDT from L2→L1 through bridge 2. This transaction fails, because bridge 2 only has 100 USDT custodied that it can give to Bob.

> The core property that this example illustrates is that locking tokens across multiple bridges on the token’s “home” domain makes it impossible to have fungibility without impeding user experience on remote domains. Minting or burning the token can solve this problem, but not all ERC-20 tokens implement a configurable mint/burn interface. 

Reviewer 2: This section doesn't seem to be explaining any particular choice(s) made within the proposal. Instead it seems like it's justifying the proposal as a whole? If so, this belongs in the Motivation section.

The Motivation should answer questions like "why do we need sovereign bridge tokens that are fungible?", while the Rationale section should answer questions like "why did we choose to include the `setLockbox` function?"

    1. This problem exists in the current model if there is insufficient liquidity on the destination for a given bridge to complete a transaction within some defined slippage constraints. This case is not well-handled by bridges at the moment. Bridges are forced to either output some wrapped representation, or force users to wait until there is more liquidity. 
    2. In the limit-based approach, the bridge will similarly not be able to complete the transaction on the destination domain until it has more capacity available. However, a limit approach provides some more reasonable guarantees to the user: (1) user have a much higher degree of predictability around time and pricing of the outputted transaction, (2) users would not receive some wrapped representation, and (3) bridges would have a simpler pathway for users to send the tokens back to the source domain or any other destination.

### Aggregation

> EIP-7281 introduces new dynamics for aggregators that massively improve user safety and experience across chains.

Reviewer 2: ```suggestion
This proposal introduces new dynamics for aggregators that massively improve user safety and experience across chains.
```

6. The issuer should at this point disallow minting *new* legacy canonical ERC-20s on the remote domain. This would organically & gradually lead to the legacy ERC-20s on the remote chain to become locked into the lockbox.
7. At any point in the future, the token issuer can then use their own treasury to unwrap xERC-20s on the remote domain into legacy ERC-20s and send them back to the home chain, incurring the latency of this on behalf of users.

### Compatibility with Bridges

One key benefit to the xERC-20 model is that the simple requirement of a burn & mint interface makes this proposal compatible with most bridges either out of the box or through a predefined custom token mapping process. The following bridges & domains were considered in the creation of this proposal:

- **3rd party bridges:** every 3rd party bridge that relies on messaging infrastructure already supports a mint and burn interface and has a pathway to map custom minted tokens.

> - **Arbitrum rollup bridge:**  Arbitrum allows token issuers to [write a custom gateway](https://developer.arbitrum.io/asset-bridging#other-flavors-of-gateways).

Reviewer 2: You'll need to remove any external links.

- **OP Stack bridge:** Optimism supports [adding a custom bridge](https://community.optimism.io/docs/guides/bridge-dev/#building-a-custom-bridge).
- **Polygon PoS bridge:** Polygon supports custom tokens [through their fx portal](https://wiki.polygon.technology/docs/develop/l1-l2-communication/fx-portal).
- **ZkSync bridge**: Zksync supports [custom bridges of any kind](https://era.zksync.io/docs/reference/concepts/bridging/bridging-asset.html#custom-bridges-on-l1-and-l2).
- **GnosisChain Omnibridge:** GnosisChain does not currently support custom tokens, but is in the process of redesigning their omnibridge and plans to allow this functionality.

Note: In most of the above cases, a custom bridge integration also means integration into canonical bridge UIs, ensuring that users have a consistent experience throughout the process. Additionally, a single xERC-20 custom bridge implementation could be built for each ecosystem and serve any number of token issuers.

## Reference Implementation

> You can find a reference implementation and associated test cases [here](https://github.com/defi-wonderland/xTokens/blob/dev/solidity/contracts/XERC-20.sol).

Reviewer 2: You'll need to either remove this section, or include a compatibly licensed (preferably CC0-1.0) reference implementation in your assets directory.
